gr-filter/lib/ival_decimator_impl.cc

```diff
@@ -0,0 +1,110 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "ival_decimator_impl.h"
+#include <gnuradio/io_signature.h>
+
+namespace gr {
+namespace filter {
+
+// Convenience defines for work function
+#define DT_IVALDECIM_BYTE 1
+#define DT_IVALDECIM_SHORT 2
+
+ival_decimator::sptr ival_decimator::make(int decimation, int data_type, int data_size)
+{
+    return gnuradio::get_initial_sptr(
+        new ival_decimator_impl(decimation, data_type, data_size));
+}
+
+/*
+ * The private constructor
+ */
+ival_decimator_impl::ival_decimator_impl(int decimation, int data_type, int data_size)
+    : gr::sync_decimator("ival_decimator",
+                         gr::io_signature::make(1, 1, data_size),
```

Comment:
Since `data_size` and `data_type` -are- directly related, let's get rid of 1 of them. either one will work alone; using both is redundant and prone to user error.

Reply:
Why wouldn't we just use data_size so it opens the block up to all data types.  Not a big fan of the new #defines.  If we go with data_type, then we should probably be able to reuse the gr enums that are exposed through swig.  With just data_size (like other blocks), could the if statement be removed and void *'s be std::copy'd two elements at a time?

Reply:
Using a `data_type` allows the `switch` and then using normal C/C++ types to do the work, which will be a -lot- more efficient than using `memcpy` or whatever. That said, the reverse could be done too: use the correct `int` type for the size provided, then do the style of copying as done already. Since this block isn't -interpreting- the data, just moving the value around, this would work.

Reply:
NOTE: of course this wouldn't work for 3 byte items, but we don't use those in GR anyway. No sure how to handle 16 byte items, since there is no `int128_t` yet TTBOMK ... but everything up to 8 byte items would be doable.